Spigot-Server-Patches/0167-Anti-Xray.patch

```diff
@@ -0,0 +1,1519 @@
++ *    for (String line : lines) {System.out.println(line);}
++ *}</code></pre>
++ * Make printLines() async:
++ * <pre><code>String[] lines;
++ *void setLine(int i, String line) { //only invoked by main thread
++ *    //lock
++ *    lines[i] = line;
++ *    //unlock
++ *}
++ *void printLines() { //only invoked by main thread
++ *    //this state of the array should be printed
++ *    new Thread() {@Override public void run() {
++ *        //here is a race condition because the main thread moved on
++ *        //and can invoke setLine()
++ *        //lock
++ *        for (String line : lines) {System.out.println(line);}
++ *        //unlock
++ *    }}.start();
++ *}</code></pre>
++ * The goal is to ensure that the correct state of the array is printed
++ * (-> the state when printLines() is invoked).
++ * As shown with the //lock and //unlock comments a normal locking system
++ * can't help here.
++ * Possible solutions:
++ * <ul>
++ * <li>Copy the array before starting the new thread and print the copy
++ * <li>Or use this Lock implementation (if setLine() is invoked infrequently)
++ * </ul>
++ * Code:
++ * <pre><code>Lock lock = new Lock();
++ *String[] lines;
++ *void setLine(int i, String line) { //only invoked by main thread
++ *    lock.waitUntilUnlock();
++ *    //no race condition possible here because
++ *    //lock.lock() is invoked on the same thread
++ *    lines[i] = line;
++ *}
++ *void printLines() { //only invoked by main thread
++ *    lock.lock();
++ *    new Thread() {@Override public void run() {
++ *        for (String line : lines) {System.out.println(line);}
++ *        lock.unlock();
++ *    }}.start();
++ *}</code></pre>
++ */
++public class Lock {
++
++    private static Lock noOpInstance = null;
++    private final AtomicInteger lockCount = new AtomicInteger();
++
++    /**
++     * Thread-safe (synchronized) singleton getter.
++     *
++     * @return The singleton instance of a no operation implementation of this class
++     */
++    public static synchronized Lock getNoOpInstance() {
++        if (noOpInstance == null) {
++            noOpInstance = new Lock() {
++                @Override
++                public void lock() {
++
++                }
++
++                @Override
++                public void unlock() {
++
++                }
++
++                @Override
++                public void waitUntilUnlock() {
++
++                }
++            };
++        }
++
++        return noOpInstance;
++    }
++
++    /**
++     * See example use case in class doc.
++     */
++    public void lock() {
++        lockCount.incrementAndGet();
++    }
++
++    /**
++     * See example use case in class doc.
++     */
++    public synchronized void unlock() {
++        lockCount.decrementAndGet();
++        notifyAll();
++    }
++
++    /**
++     * See example use case in class doc.
++     */
++    public synchronized void waitUntilUnlock() {
++        try {
++            while (lockCount.get() > 0) {
```

Comment:
This section doesn't look right. This isn't the right pattern for (test and wait if failed) pattern.

The try/catch should be inside of the while and ignore the interrupt.